discv5/discv5-theory.md

```diff
@@ -41,6 +40,141 @@ IP address and port.
+
+    A -> B   FINDNODE
+    A -> B   PING
+    A -> B   TOPICQUERY
+    A <- B   WHOAREYOU (token references PING)
+
+In those cases, pending requests can be considered invalid (the remote end cannot decrypt
+them) and the packet referenced by WHOAREYOU (example: PING) must be re-sent with an
+authentication header. When the response to the re-sent request (example: PONG) is
+received, the new session is established and other pending requests (example: FINDNODE,
+TOPICQUERY) may be re-sent.
+
+Note that WHOAREYOU is only ever valid as a response to a previously sent request. If
+WHOAREYOU is received but no requests are pending, the handshake attempt can be ignored.
+
+### Identity-Specific Cryptography in the Handshake
+
+Establishment of session keys is dependent on the identity scheme of the recipient (i.e.
+the node which sends WHOAREYOU). Similarly, the signature over `id-nonce-input` is made by
+the identity key of the initiator. Although initiator and recipient might not be using the
+same identity scheme in their respective node records, implementations must be able to
+handle handshaking for all supported identity schemes.
+
+At this time, the only supported identity scheme is "v4".
+
+`id_sign(data)` creates a signature over `data` using the node's static private key. The
+signature is encoded as the 64-byte array `r || s`, i.e. as the concatenation of the
+signature values.
+
+`ecdh(pubkey, privkey)` creates a secret through elliptic-curve Diffie-Hellman key
+agreement. The public key is multiplied by the private key to create a secret ephemeral
+key `eph = pubkey * privkey`. The 33-byte secret output is `y || eph.x` where `y` is
+`0x02` when `eph.y` is even or `0x03` when `eph.y` is odd.
+
+### Session Cache
+
+Nodes should store session keys for communication with other recently-seen nodes. Since
+sessions are ephemeral and can be re-established whenever necessary, it is sufficient to
+store a limited number of sessions in an in-memory LRU cache.
+
+To prevent IP spoofing attacks, implementations must ensure that session secrets and the
+handshake are tied to a specific UDP endpoint. This is simple to implement by using the
+node ID and IP/port as the 'key' into the in-memory session cache. When a node switches
+endpoints, e.g. when roaming between different wireless networks, sessions will to be
+re-established by handshaking again. This requires no effort on behalf of the roaming node
+because the recipients of protocol messages will simply refuse to decrypt messages from
+the new endpoint and reply with WHOAREYOU.
+
+The number of messages which can be encrypted with a certain session key is limited
+because encryption of each message requires a unique nonce for AES-GCM. In addition to the
+keys, the session cache must also keep track of the count of outgoing messages to ensure
+the uniqueness of nonce values. Since the wire protocol uses 96 bit AES-GCM nonces, it is
```

Comment:
According to [this recommendation, section 8.2.2](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf#%5B%7B%22num%22%3A65%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C0%2C792%2Cnull%5D), it would be secure to get the first `IV` from a randomly generated bit string and then increment its e.g. 32-bit part to derive values for next `IVs`.

Reply:
Yes, that would work as well.

Reply:
Sure, I was also trying to say that 32 bit counter + 64 random bits is not considered as a secure option. It might be worth changing this sentence to what we have in the recommendation.